lib/rules/utils/char-source.js

```diff
@@ -0,0 +1,226 @@
+/**
+ * @fileoverview Utility functions to locate the source text of each code unit in the value of a string literal or template token.
+ * @author Francesco Trotta
+ */
+
+"use strict";
+
+/**
+ * Represents a code unit produced by the evaluation of a JavaScript common token like a string
+ * literal or template token.
+ */
+class CodeUnit {
+    constructor(start, source) {
+        this.start = start;
+        this.source = source;
+    }
+
+    get end() {
+        return this.start + this.length;
+    }
+
+    get length() {
+        return this.source.length;
+    }
+}
+
+/**
+ * An object used to keep track of the position in a source text where the next characters will be read.
+ */
+class SourceReader {
```

Comment:
I might rename this as `TextSource`. `SourceReader` implies that there's a `read()` method or something that this object is supposed to be doing, but it's just a data structure.

Reply:
Thanks for the feedback. I've added a `read()` method in https://github.com/eslint/eslint/pull/18082/commits/448d9a9e40e8f001b59352a730098658f7d55081 to read characters relative to the current position. So it's no longer necessary to store the position in a variable before accessing the source like it was done before. And I'm using `+=` for relative position changes. I think this makes the code a little clearer but if that's not the case we can revert to how it was done previously and just address the suggestions.

Reply:
I like the `read()`!

`SourceReader` is still a little ambiguous IMO. Maybe, `CharsReader`? `CharactersReader`? `TestReader`? Not a blocker from my end.

Reply:
Thanks! I don't have a preference for the name, so pick one that sounds better to you.

Reply:
I'd vote for `TextReader` then, to align with https://github.com/eslint/eslint/pull/18082#discussion_r1484649381.

Reply:
Renamed in 072f256d21e21551f896621312578332535fbef3.